src/mame/tandberg/tdv2100_kbd.cpp

```diff
@@ -0,0 +1,577 @@
+							break;
+
+						case 14:
+							switch(row_counter)
+							{
+								case 4:
+									m_write_cleark_cb(1);
+									break;
+
+								case 5:
+									if(!m_8_bit_output)
+									{
+										m_write_transk_cb(1);
+									}
+									break;
+
+								case 6:
+									m_write_linek_cb(1);
+									break;
+
+								case 7:
+									m_write_break_cb(1);
+									break;
+							}
+							break;
+
+						default:
+							new_keystroke(key_nr, (m_shift||m_shift_lock), m_control);
+					}
+				}
+				else
+				{
+					switch(m_column_counter)
+					{
+						case 0:
+							switch(row_counter)
+							{
+								case 0:
+								case 1:
+									m_keystate[m_column_counter] &= ~(1<<row_counter);  // Turn off released shift key
+									m_shift = m_keystate[m_column_counter]&0x03;	    // Keep shift state if the other shift key is still down
+									break;
+
+								case 3:
+									m_control = false;
+									break;
+							}
+							break;
+
+						case 14:
+							switch(row_counter)
+							{
+								case 4:
+									m_write_cleark_cb(0);
+									break;
+
+								case 5:
+									if(!m_8_bit_output)
+									{
+										m_write_transk_cb(0);
+									}
+									break;
+
+								case 6:
+									m_write_linek_cb(0);
+									break;
+
+								case 7:
+									m_write_break_cb(0);
+									break;
+							}
+							break;
+
+						default:
+							if(key_nr == m_key_nr_in_buffer)
+							{
+								m_key_repeat_trigger->adjust(attotime::never);
+							}
+					}
+				}
+			}
+		}
+		m_keystate[m_column_counter] = column;
+
+		m_column_counter++;
+		if(m_column_counter >= 15)
+		{
+			m_column_counter = 0;
+		}
+	}
+}
+
+void tandberg_tdv2100_keyboard_device::new_keystroke(uint8_t key_nr, bool shift, bool control)
+{
+	uint8_t param = m_keyparams[(key_nr&0x7f) | ((control) ? 0x80 : 0x00)];
+	bool all_caps = !(m_sw_all_cap->read());
+	shift = (shift || ((param&0x01) && all_caps));
+	uint8_t key = m_keymap[(key_nr&0x07f) | ((shift) ? 0x080 : 0x000) | ((control) ? 0x100 : 0x000)];
+
+	bool inhibit = (key&0x80);
+	if(m_8_bit_output)
+	{
+		inhibit = (param&0x04);
+	}
+
+	if(!inhibit)
+	{
+		m_key_nr_in_buffer = key_nr;
+		m_char_buffer = key&0x7f;
+		if(m_8_bit_output)
+		{
+			m_write_transk_cb((key>>7)&0x01);
+		}
+		key_trigger();
+		if(param&0x02)
+		{
+			int key_repeat_delay_ms = 500 + m_key_repeat_delay->read()*750/100;    // 500ms -> 1250ms
+			m_key_repeat_trigger->adjust(attotime::from_msec(key_repeat_delay_ms));
+		}
+	}
+}
+
+TIMER_CALLBACK_MEMBER(tandberg_tdv2100_keyboard_device::key_repeat)
+{
+	key_trigger();
+	int key_repeat_rate_hz = 10 + m_key_repeat_rate->read()/5;    // 10Hz -> 30Hz
+	m_key_repeat_trigger->adjust(attotime::from_hz(key_repeat_rate_hz));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Strobes
+//
+
+void tandberg_tdv2100_keyboard_device::key_trigger()
+{
+	m_write_kstr_cb(0x80|m_char_buffer);
+	m_write_kstr_cb(m_char_buffer);
```

Comment:
Is there potential for issues to be caused by the lack of any time between these two calls?  How long would the top bit be asserted for in reality?